File: lib/config/loader.ts
// Loads the merged config.yaml into a typed AppConfig. A missing section yields
// a sensible empty default (e.g. no products: section -> { entries: [] }).

import { parse as parseYaml } from "@std/yaml";
import type { AppConfig } from "./types.ts";

interface RawConfig {
  types?: AppConfig["types"];
  field_descriptions?: Record<string, string>;
  products?: AppConfig["products"];
}

export async function loadConfig(path = "config.yaml"): Promise<AppConfig> {
  const doc = (parseYaml(await Deno.readTextFile(path)) as RawConfig | null) ?? {};
  return {
    types: doc.types ?? {},
    fieldDescriptions: doc.field_descriptions ?? {},
    products: doc.products ?? { entries: [] },
  };
}

// Optional override file for field descriptions (the deprecated --fields-doc flag).
// Accepts either { descriptions: {...} } (legacy field_descriptions.yaml) or a
// bare map. Returns {} if absent/unparseable.
export async function loadFieldDescriptionsFile(path: string): Promise<Record<string, string>> {
  try {
    const doc = parseYaml(await Deno.readTextFile(path)) as
      | { descriptions?: Record<string, string> }
      | Record<string, string>
      | null;
    if (!doc || typeof doc !== "object") return {};
    const maybe = (doc as { descriptions?: Record<string, string> }).descriptions;
    return (maybe && typeof maybe === "object" ? maybe : (doc as Record<string, string>)) ?? {};
  } catch {
    return {};
  }
}

File: f5kb/config/loader.py
"""Loads the merged config.yaml into a typed AppConfig."""

from __future__ import annotations

from pathlib import Path

import yaml

from f5kb.config.types import AppConfig, ProductEntry, ProductsSection, normalize_type


def load_config(path: str = "config.yaml") -> AppConfig:
    text = Path(path).read_text(encoding="utf-8")
    doc = yaml.safe_load(text) or {}

    raw_types = doc.get("types") or {}
    types = {k: normalize_type(v or {}) for k, v in raw_types.items()}

    field_descriptions: dict[str, str] = doc.get("field_descriptions") or {}

    raw_products = doc.get("products") or {}
    raw_entries = raw_products.get("entries") or []
    entries = tuple(
        ProductEntry(
            product=e.get("product") or "",
            count=int(e.get("count") or 0),
            source=e.get("source") or "",
            hidden_from_global_facet=bool(e.get("hiddenFromGlobalFacet") or False),
            discovered_via_types=tuple(e.get("discoveredViaTypes") or []),
        )
        for e in raw_entries
    )
    products = ProductsSection(
        generated_at=raw_products.get("generatedAt"),
        entries=entries,
    )

    return AppConfig(types=types, field_descriptions=field_descriptions, products=products)


def types_for_lambda(path: str = "config.yaml") -> dict[str, dict]:
    """Serialise the `types:` block for s3://<bucket>/lambda/config/types.json.

    Shape the Lambdas consume: {type_key: {documentType, metadata, content}}.
    The Dump Lambda reads per-type entries (field split + the exact Coveo
    documentType filter value); the Orchestrator's legacy branch derives a
    type list from the mapping keys when TYPE_KEYS is unset.
    """
    cfg = load_config(path)
    return {
        key: {
            "documentType": tc.document_type,
            "metadata": tc.metadata if tc.metadata == "*" else list(tc.metadata),
            "content": tc.content if tc.content == "*" else list(tc.content),
        }
        for key, tc in cfg.types.items()
    }


def load_field_descriptions_file(path: str) -> dict[str, str]:
    """Optional override file for field descriptions. Returns {} on any error."""
    try:
        text = Path(path).read_text(encoding="utf-8")
        doc = yaml.safe_load(text)
        if not doc or not isinstance(doc, dict):
            return {}
        maybe = doc.get("descriptions")
        if isinstance(maybe, dict):
            return maybe
        return doc
    except Exception:
        return {}
